### PR TITLE
Add AbsFOp, CeilFOp, CosOp, NegFOp, SignFOp, TanhOp to Standard dialect.

### PR DESCRIPTION
**Pull request by dfki-jugr**
Include documentation and tests for the newly added operations.

### CODE REVIEW

File: g3doc/Dialects/Standard.md [hunk 1]
```diff
@@ -479,6 +479,156 @@ The `exp` operation takes one operand and returns one result of the same type.
 This type may be a float scalar type, a vector whose element type is float, or a
 tensor of floats. It has no standard attributes.
 
+### 'abs' operation
```
Comment (sherhut) [inline g3doc/Dialects/Standard.md]: Please keep these sorted.

File: g3doc/Dialects/Standard.md [hunk 2]
```diff
@@ -479,6 +479,156 @@ The `exp` operation takes one operand and returns one result of the same type.
...
+Syntax:
+
+``` {.ebnf}
+operation ::= ssa-id `=` `tanh` ssa-use `:` type
+```
+
+Examples:
+
+```mlir {.mlir}
+// Scalar hyperbolic tangents value.
```
Comment (sherhut) [inline g3doc/Dialects/Standard.md]: hyperbolic tangent?

File: include/mlir/Dialect/StandardOps/Ops.td [hunk 3]
```diff
@@ -133,6 +133,10 @@ class FloatArithmeticOp<string mnemonic, list<OpTrait> traits = []> :
     ArithmeticOp<mnemonic, traits>,
     Arguments<(ins FloatLike:$lhs, FloatLike:$rhs)>;
 
+def FAbsOp : FloatUnaryOp<"fabs"> {
```
Comment (sherhut) [inline include/mlir/Dialect/StandardOps/Ops.td]: Keep in style with the other operations, which have the F in the end.

File: include/mlir/Dialect/StandardOps/Ops.td [hunk 4]
```diff
@@ -950,6 +966,10 @@ def SelectOp : Std_Op<"select", [NoSideEffect, SameOperandsAndResultShape]> {
   let hasFolder = 1;
 }
 
+def SignOp : FloatUnaryOp<"sign"> {
```
Comment (sherhut) [inline include/mlir/Dialect/StandardOps/Ops.td]: SignFOp, as this is limited to floats and a SignIOp would make sense. If it is SignOp, I would prefer it to work on both.

File: test/IR/core-ops.mlir [hunk 5]
```diff
@@ -370,6 +370,51 @@ func @standard_instrs(tensor<4x4x?xf32>, f32, i32, index, i64, f16) {
 
   // CHECK: %{{[0-9]+}} = exp %arg0 : tensor<4x4x?xf32>
   %98 = exp %t : tensor<4x4x?xf32>
+
```
Comment (sherhut) [inline test/IR/core-ops.mlir]: This is missing tests for the vector variants.

File: include/mlir/Dialect/StandardOps/Ops.td [hunk 6]
```diff
@@ -855,6 +867,10 @@ def MulIOp : IntArithmeticOp<"muli", [Commutative]> {
   let hasFolder = 1;
 }
 
+def FNegOp : FloatUnaryOp<"fneg"> {
```
Comment (pifon2a) [inline include/mlir/Dialect/StandardOps/Ops.td]: NegFOp?

File: include/mlir/Dialect/StandardOps/Ops.td [hunk 7]
```diff
@@ -345,6 +349,10 @@ def CallIndirectOp : Std_Op<"call_indirect", [CallOpInterface]> {
   let hasCanonicalizer = 1;
 }
 
+def CeilOp : FloatUnaryOp<"ceil"> {
```
Comment (pifon2a) [inline include/mlir/Dialect/StandardOps/Ops.td]: CeilFOp? for same reasons as for SignFOp, as Stephan mentioned.

File: include/mlir/Dialect/StandardOps/Ops.td [hunk 8]
```diff
@@ -601,6 +609,10 @@ def ConstantOp : Std_Op<"constant", [NoSideEffect]> {
   let hasFolder = 1;
 }
 
+def CosOp : FloatUnaryOp<"cos"> {
+  let summary = "cosine value of the specified value";
```
Comment (pifon2a) [inline include/mlir/Dialect/StandardOps/Ops.td]: nit: "cosine of the specified value"

File: include/mlir/Dialect/StandardOps/Ops.td [hunk 9]
```diff
@@ -345,6 +349,10 @@ def CallIndirectOp : Std_Op<"call_indirect", [CallOpInterface]> {
   let hasCanonicalizer = 1;
 }
 
+def CeilOp : FloatUnaryOp<"ceil"> {
+  let summary = "ceiling value of the specified value";
```
Comment (pifon2a) [inline include/mlir/Dialect/StandardOps/Ops.td]: nit: "ceiling of the specified value"

File: include/mlir/Dialect/StandardOps/Ops.td [hunk 10]
```diff
@@ -855,6 +867,10 @@ def MulIOp : IntArithmeticOp<"muli", [Commutative]> {
   let hasFolder = 1;
 }
 
+def FNegOp : FloatUnaryOp<"fneg"> {
+  let summary = "floating point negate operation";
```
Comment (pifon2a) [inline include/mlir/Dialect/StandardOps/Ops.td]: nit: "floating point negation"

File: include/mlir/Dialect/StandardOps/Ops.td [hunk 11]
```diff
@@ -950,6 +966,10 @@ def SelectOp : Std_Op<"select", [NoSideEffect, SameOperandsAndResultShape]> {
   let hasFolder = 1;
 }
 
+def SignOp : FloatUnaryOp<"sign"> {
+  let summary = "floating point negate operation";
```
Comment (pifon2a) [inline include/mlir/Dialect/StandardOps/Ops.td]: please update the comment. This one seem to be copied from `Neg`.

File: include/mlir/Dialect/StandardOps/Ops.td [hunk 12]
```diff
@@ -1064,6 +1084,10 @@ def StoreOp : Std_Op<"store"> {
   let hasCanonicalizer = 1;
 }
 
+def TanhOp : FloatUnaryOp<"tanh"> {
+  let summary = "hyperbolic tangents value of the specified value";
```
Comment (pifon2a) [inline include/mlir/Dialect/StandardOps/Ops.td]: nit: "hyperbolic tangent of the specified value"

File: test/IR/core-ops.mlir [hunk 13]
```diff
@@ -370,6 +370,51 @@ func @standard_instrs(tensor<4x4x?xf32>, f32, i32, index, i64, f16) {
...
+  // CHECK: %{{[0-9]+}} = fabs %arg1 : f32
+  %99 = "std.fabs"(%f) : (f32) -> f32
+
```
Comment (pifon2a) [inline test/IR/core-ops.mlir]: i am not sure if it should be `fabs` or `absf` analoguos to `addf`.

File: include/mlir/Dialect/StandardOps/Ops.td (hunk 12 above)
Comment (pifon2a) [reply include/mlir/Dialect/StandardOps/Ops.td]: PTAL

File: g3doc/Dialects/Standard.md [hunk 14]
```diff
@@ -454,6 +454,31 @@ tensor_store %8, %10 : memref<4x?xf32, #layout, memspace0>
 
 ## Unary Operations
 
+### 'abs' operation
```
Comment (pifon2a) [inline g3doc/Dialects/Standard.md]: `%s/abs/absf/gc`

File: g3doc/Dialects/Standard.md [hunk 15]
```diff
@@ -479,6 +504,131 @@ The `exp` operation takes one operand and returns one result of the same type.
 This type may be a float scalar type, a vector whose element type is float, or a
 tensor of floats. It has no standard attributes.
 
+### 'ceil' operation
```
Comment (pifon2a) [inline g3doc/Dialects/Standard.md]: `%s/ceil/ceilf/gc`

File: g3doc/Dialects/Standard.md [hunk 16]
```diff
@@ -479,6 +504,131 @@ The `exp` operation takes one operand and returns one result of the same type.
...
+
+// Tensor element-wise cosine value.
+%x = cos %y : tensor<4x?xf8>
+```
+
+The `cos` operation takes one operand and returns one result of the same type.
+This type may be a float scalar type, a vector whose element type is float, or a
+tensor of floats. It has no standard attributes.
+
+### 'neg' operation
```
Comment (pifon2a) [inline g3doc/Dialects/Standard.md]: `%s/neg/negf/gc`

File: g3doc/Dialects/Standard.md [hunk 17]
```diff
@@ -479,6 +504,131 @@ The `exp` operation takes one operand and returns one result of the same type.
...
+
+// Tensor element-wise negate value.
+%x = neg %y : tensor<4x?xf8>
+```
+
+The `neg` operation takes one operand and returns one result of the same type.
+This type may be a float scalar type, a vector whose element type is float, or a
+tensor of floats. It has no standard attributes.
+
+### 'sign' operation
```
Comment (pifon2a) [inline g3doc/Dialects/Standard.md]: `%s/sign/signf/gc`

File: test/IR/core-ops.mlir [hunk 18]
```diff
@@ -368,8 +368,71 @@ func @standard_instrs(tensor<4x4x?xf32>, f32, i32, index, i64, f16) {
...
+  // CHECK: %{{[0-9]+}} = negf %arg1 : f32
+  %112 = "std.negf"(%f) : (f32) -> f32
+
+  // CHECK: %{{[0-9]+}} = negf %arg1 : f32
+  %113 = negf %f : f32
+
+  // CHECK: %{{[0-9]+}} = negf %arg1 : vector<4xf32>
+  %114 = negf %v : vector<4xf32>
+
```
Comment (pifon2a) [inline test/IR/core-ops.mlir]: I do not see the tests for `signf`.

File: include/mlir/Dialect/StandardOps/Ops.td [hunk 19]
```diff
@@ -950,6 +966,14 @@ def SelectOp : Std_Op<"select", [NoSideEffect, SameOperandsAndResultShape]> {
   let hasFolder = 1;
 }
 
+def SignFOp : FloatUnaryOp<"signf"> {
+  let summary = "floating point sign operation";
+}
+
+def SignIOp : IntegerUnaryOp<"signi"> {
```
Comment (pifon2a) [inline include/mlir/Dialect/StandardOps/Ops.td]: Remove SignIOp?

File: g3doc/Dialects/Standard.md [hunk 20]
```diff
@@ -479,6 +504,131 @@ The `exp` operation takes one operand and returns one result of the same type.
...
+%f = tanh %g : vector<4xf32>
+
+// Tensor element-wise hyperbolic tangent value.
+%x = tanh %y : tensor<4x?xf8>
+```
+
+The `tanh` operation takes one operand and returns one result of the same type.
+This type may be a float scalar type, a vector whose element type is float, or a
+tensor of floats. It has no standard attributes.
+
```
Comment (joker-eph) [inline g3doc/Dialects/Standard.md]: Ultimately we should rework all this to have the dialect documentation be generated from the ODS definition. Can you make sure at least that all the information here is present in the ODS file?

Comment (pifon2a) [reply g3doc/Dialects/Standard.md]: I will add description fields for the rest of the ops in the follow-up PR. If the description is missing, of course.

File: g3doc/Dialects/Standard.md [hunk 21]
```diff
@@ -454,6 +454,31 @@ tensor_store %8, %10 : memref<4x?xf32, #layout, memspace0>
...
+// SIMD vector element-wise absolute value.
+%f = absf %g : vector<4xf32>
+
+// Tensor element-wise absolute value.
+%x = absf %y : tensor<4x?xf8>
+```
+
+The `absf` operation takes one operand and returns one result of the same type.
+This type may be a float scalar type, a vector whose element type is float, or a
+tensor of floats. It has no standard attributes.
```
Comment (joker-eph) [inline g3doc/Dialects/Standard.md]: Can you please describe the operations more than just structurally?

File: include/mlir/Dialect/StandardOps/Ops.td [hunk 22]
```diff
@@ -950,6 +986,15 @@ def SelectOp : Std_Op<"select", [NoSideEffect, SameOperandsAndResultShape]> {
...
 
+def SignFOp : FloatUnaryOp<"signf"> {
+  let summary = "floating point sign operation";
+  let description = [{
+    The `signf` operation takes one operand and returns one result of the same type.
+    This type may be a float scalar type, a vector whose element type is float, or a
+    tensor of floats. It has no standard attributes.
+  }]
+}
+
```
Comment (joker-eph) [inline include/mlir/Dialect/StandardOps/Ops.td]: Where are "signf" and "negf" coming from? What motivates adding these to the standard dialect?
(for instance I don't think LLVM has these).

Comment (pifon2a) [reply include/mlir/Dialect/StandardOps/Ops.td]: They will be lowered to copysign when we do LLVM legalization.

We can replace these ops with the use of `copysign`. Then the lowering from LHLO will be a bit longer, but STD->LLVM will be trivial. What do you prefer? Is the point of Standard dialect to be as close to LLVM, but still allow to lower to LLVM, Cuda, Rocm if needed?

File: g3doc/Dialects/Standard.md (hunk 21 above)
Comment (pifon2a) [reply g3doc/Dialects/Standard.md]: +1

File: include/mlir/Dialect/StandardOps/Ops.td (hunk 22 above)
Comment (joker-eph) [reply include/mlir/Dialect/StandardOps/Ops.td]: > Is the point of Standard dialect to be as close to LLVM, but still allow to lower to LLVM, Cuda, Rocm if needed?

Standard dialect does not really have a well defined purpose, but it is central enough that we should have a higher level of scrutiny and consideration about what is going there.

Can you send an email to mlir@tensorflow.org to propose these, explaining their relationship to `copysign`?

File: include/mlir/Dialect/StandardOps/Ops.td [hunk 23]
```diff
@@ -136,10 +136,10 @@ class FloatArithmeticOp<string mnemonic, list<OpTrait> traits = []> :
 def AbsFOp : FloatUnaryOp<"absf"> {
   let summary = "floating point absolute-value operation";
   let description = [{
-    The `absf` operation takes one operand and returns one result of the same type.
-    This type may be a float scalar type, a vector whose element type is float, or a
-    tensor of floats. It has no standard attributes.
-  }]
+    The `absf` operation takes one operand and returns one result of the same
```
Comment (pifon2a) [inline include/mlir/Dialect/StandardOps/Ops.td]: Please, copy the updated documentation from Standard.md here.

**Pull request closed without merging**